### PR TITLE
gh-91378: Allow subprocess pass-thru with stdout/stderr capture

### PR DESCRIPTION
**Pull request by pprindeville**
# Allow pass-thru of subprocess output even when capturing to buffers

I maintain some build wrappers that both pass the output (both `stdout` and `stderr`) through when being run interactively, but also capture logs into artifacts when being run through a CI/CD pipeline.

Being able to call `proc = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, tee=True)` would allow harvesting results as `output, errors = subprocess.communicate()` as well as permitting the results of the running process to be seen in real-time (some `make` recipes invoke `docker` and can take a while to complete... buffering until completion would confuse users as it might appear that the recipe has hung). 

https://github.com/python/cpython/issues/91378, formerly:

https://bugs.python.org/issue47222

* Issue: gh-91378

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@pprindeville

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by pprindeville** [comment]
@Andrew-Shay This should look familiar...

File: Lib/subprocess.py [hunk 1]
```diff
@@ -1502,7 +1504,13 @@ def _wait(self, timeout):
 
 
         def _readerthread(self, fh, buffer):
-            buffer.append(fh.read())
+            data = fh.read()
+            if self.tee:
+                if fh == self.stdout:
+                    sys.stdout.write(data.decode())
+                elif fh == self.stderr:
+                    sys.stderr.write(data.decode())
```
Comment (gst) [inline Lib/subprocess.py]:
```python
if fh == self.stdout or fh == self.stderr:
    fh.write(data.decode())
```

?

**Review by gpshead — CHANGES_REQUESTED** [review]
A few things:

 1. [ ] You're calling `data.decode()` on the read data, but no codec is specified. This is going to cause problems as there is never a guarantee about what an arbitrary process will actually output. A true "tee" should take the binary data it read and write the same binary data out to the destination. Which means you should use `sys.stdout.buffer.write(data)` rather than `sys.stdout.write(data.decode())` (same for stderr).
 2. [ ] This needs a unittest. Look around `test_subprocess.py` for ideas. Executing a child process that executes a child process to see that the expected output winds up in the right place is entirely within reason.
 3. [ ] What happens when `tee` is True but the stdout and/or stderr were not `PIPE`? If this would be useless, we should make it an error.
 4. [ ] Documentation and docstrings need updating (I usually leave this to last), with a practical example.

I'm not decided on if I like this feature as designed or not yet, but I've seen enough code that could use it to know that the need is real in some form or another (you gave a good practical example). So lets get this into shape and see how it looks before making a final decision if this is the desired API or should be tweaked into a different shape.

**Comment by pprindeville** [comment]
> A few things:

@gpshead Thanks for the comments.

> 1. [ ]  You're calling `data.decode()` on the read data, but no codec is specified. This is going to cause problems as there is never a guarantee about what an arbitrary process will actually output. A true "tee" should take the binary data it read and write the same binary data out to the destination. Which means you should use `sys.stdout.buffer.write(data)` rather than `sys.stdout.write(data.decode())` (same for stderr).

Done

> 2. [ ]  This needs a unittest. Look around `test_subprocess.py` for ideas. Executing a child process that executes a child process to see that the expected output winds up in the right place is entirely within reason.

Took a stab at this, but I'm unfamiliar with the test framework or how to debug the failure... any suggestions are appreciated.

> 3. [ ]  What happens when `tee` is True but the stdout and/or stderr were not `PIPE`? If this would be useless, we should make it an error.

Added a test.  Please let me know if you feel it's adequate or not.

> 4. [ ]  Documentation and docstrings need updating (I usually leave this to last), with a practical example.
> 
> I'm not decided on if I like this feature as designed or not yet, but I've seen enough code that could use it to know that the need is real in some form or another (you gave a good practical example). So lets get this into shape and see how it looks before making a final decision if this is the desired API or should be tweaked into a different shape.

Agreed.  I'll do this last.

**Comment by pprindeville** [comment]
Update:

> 2. [ ]  This needs a unittest. Look around `test_subprocess.py` for ideas. Executing a child process that executes a child process to see that the expected output winds up in the right place is entirely within reason.

@gpshead Added 2 unit tests.

**Comment by pprindeville** [comment]
I have made the requested changes; please review again

File: Lib/subprocess.py [hunk 2]
```diff
@@ -756,19 +756,34 @@ class Popen:
...
+
+      tee: If true, the output and/or error streams are also written to
+          the parent's sys.stdout and/or sys.stderr, respectively.
+
     Attributes:
         stdin, stdout, stderr, pid, returncode
     """
     _child_created = False  # Set here since __del__ checks it
 
+    """Default handler for read_cb."""
```
Comment (gpshead) [inline Lib/subprocess.py]: docstrings belong inside the def.

File: Doc/library/subprocess.rst [hunk 3]
```diff
@@ -654,6 +655,16 @@ functions.
    .. versionadded:: 3.10
       The ``pipesize`` parameter was added.
 
+   *read_cb* can be used to replace the default handler for reading the
```
Comment (gpshead) [inline Doc/library/subprocess.rst]: "to replace the default _internal_ handler for"

and strike the ", and is subprocess.Popen._read_handler" text.  we don't need to tell people what the internal-only thing is in the public API docs.  just that they can replace it.

Describe the required callable signature in the docs here and how it is expected to behave.  Don't force people to read subprocess.py in order to successfully implement a read_handler.

File: Lib/subprocess.py [hunk 4]
```diff
@@ -815,6 +836,13 @@ def __init__(self, args, bufsize=-1, executable=None,
         self.encoding = encoding
         self.errors = errors
         self.pipesize = pipesize
+        if read_cb is not None:
+            if not callable(read_cb):
+                ValueError("read_cb not a callable")
+            self.read_cb = read_cb
+        else:
+            self.read_cb = Popen._read_handler
```
Comment (gpshead) [inline Lib/subprocess.py]: `self._read_handler` would be better.  it is unusual to see code call `self.foo(self, ...)` which is what the current way of implementing it in this PR revision does.  that makes for hard to follow code.  I realize you're doing it because you want to allow injection of a new read_cb via the constructor which doesn't have an instance to access yet, but this makes for an awkward API.  it doesn't need self, per my other comment just pass an indicator of if the call is being made for stdout or for stderr data.  Or have a separate read_callback parameter for each instead so that conditional logic need not be duplicated within callback implementations.

File: Doc/library/subprocess.rst [hunk 5]
```diff
@@ -342,7 +342,8 @@ functions.
                  startupinfo=None, creationflags=0, restore_signals=True, \
                  start_new_session=False, pass_fds=(), *, group=None, \
                  extra_groups=None, user=None, umask=-1, \
-                 encoding=None, errors=None, text=None, pipesize=-1)
+                 encoding=None, errors=None, text=None, pipesize=-1, \
+                 read_cb=None, tee=False)
```
Comment (gpshead) [inline Doc/library/subprocess.rst]: I suggest avoiding the abbreviation, call it `read_callback=` if this is going to exist rather than be a method that a subclass of POpen could override.

File: Lib/test/test_subprocess.py [hunk 6]
```diff
@@ -316,26 +318,112 @@ def test_stdout_none(self):
...
+        # .stdout is PIPE when not redirected, and the child's stdout will
+        # be inherited from the parent.  In order to test this we run a
+        # subprocess in a subprocess:
+        # this_test
+        #   \-- subprocess created by this test (parent)
+        #          \-- subprocess created by the parent subprocess (child)
+        # The parent doesn't specify stdout, so the child will use the
+        # parent's stdout.  This test checks that the message printed by the
+        # child goes to the parent stdout.  The parent also checks that the
+        # child's stdout is cloned.  See #47222.
```
Comment (gpshead) [inline Lib/test/test_subprocess.py:353]: lets refer to gh-91378 instead of the old bugs.python.org number now.

File: Lib/subprocess.py [hunk 7]
```diff
@@ -756,19 +756,34 @@ class Popen:
...
+      tee: If true, the output and/or error streams are also written to
+          the parent's sys.stdout and/or sys.stderr, respectively.
+
     Attributes:
         stdin, stdout, stderr, pid, returncode
     """
     _child_created = False  # Set here since __del__ checks it
 
+    """Default handler for read_cb."""
+    def _read_handler(self, fh, buffer, data):
```
Comment (gpshead) [inline Lib/subprocess.py]: Rather than passing fh to the read handler, just a simple value indicating if it is stdout or stderr would be good. With that there is no need for self at all and this can become a `@staticmethod` rather and the whole `self` as a parameter issue disappears.

**Review by gpshead — CHANGES_REQUESTED** [review]
[theme: still iterating on the design]

While I still question if we should really add the callbacks at all, we probably should: For people who want that concept perhaps they could maybe use asyncio instead https://docs.python.org/3/library/asyncio-subprocess.html#examples to await data from the pipes?  But that isn't necessarily a great API either and has its own drawbacks (requiring understanding asyncio concepts at all)...

File: Lib/subprocess.py (hunk 7 above)
Comment (pprindeville) [reply Lib/subprocess.py]: If someone subclasses `Popen` and adds further state that they want to access, how will they do that?  And we could pass in a bool like `is_stdout`, but we wouldn't have access to `self.tee` which we still need.

File: Lib/test/test_subprocess.py (hunk 6 above)
Comment (pprindeville) [reply Lib/test/test_subprocess.py:353]: Done

File: Doc/library/subprocess.rst (hunk 5 above)
Comment (pprindeville) [reply Doc/library/subprocess.rst]: Done

File: Lib/subprocess.py (hunk 4 above)
Comment (pprindeville) [reply Lib/subprocess.py]: > it doesn't need self, per my other comment just pass an indicator of if the call is being made for stdout or for stderr data. Or have a separate read_callback parameter for each instead so that conditional logic need not be duplicated within callback implementations.

If someone derives from `Popen` and adds additional internal state, we can't know what the handler will need access to and what it won't.  The only way around this is giving the handler access to `self`.

File: Lib/subprocess.py (hunk 2 above)
Comment (pprindeville) [reply Lib/subprocess.py]: Done

File: Lib/subprocess.py (hunk 7 above)
Comment (pprindeville) [reply Lib/subprocess.py]: Changed it to take `is_stdout` (bool) instead of `fh`.

**Comment by pprindeville** [comment]
I've made all the changes discussed except one which doesn't account for subclassing.  I think that's a consideration that needs to be addressed, so I'm not going to make _all_ the changes just so I can claim, "I have made the requested changes; please review again".

**Comment by pprindeville** [comment]
Can I please get this re-reviewed?

File: Lib/subprocess.py [hunk 8]
```diff
@@ -815,6 +839,13 @@ def __init__(self, args, bufsize=-1, executable=None,
         self.encoding = encoding
         self.errors = errors
         self.pipesize = pipesize
+        if read_callback is not None:
+            if not callable(read_callback):
+                ValueError("read_callback not a callable")
```
Comment (gst) [inline Lib/subprocess.py]: missing raise and associated test case thus ;)

**Review by gst — CHANGES_REQUESTED** [review]

File: Lib/subprocess.py (hunk 8 above)
Comment (pprindeville) [reply Lib/subprocess.py]: > missing raise and associated test case thus ;)

Done

File: Lib/subprocess.py [hunk 9]
```diff
@@ -1502,7 +1533,9 @@ def _wait(self, timeout):
 
 
         def _readerthread(self, fh, buffer):
-            buffer.append(fh.read())
+            data = fh.read()
```
Comment (gpshead) [inline Lib/subprocess.py:1598]: Windows uses threads for the stdout and stderr reading so that it can just do this single blocking unbounded blocking read(). Keeping this behavior when a read_callaback or tee is specified seems to defeat the purpose as it'll wait until the process has completed before calling the callback (tee'ing the output) all at once. That isn't what someone who has specified tee wants and probably isn't what someone who has specified their own read_callback wants. As there is little point in either of those if you don't get data asynchronously as it comes in.

File: Lib/subprocess.py [hunk 10]
```diff
@@ -756,19 +756,37 @@ class Popen:
       encoding and errors: Text mode encoding and error handling to use for
           file objects stdin, stdout and stderr.
 
+      read_callback: Callable to handle reading from the output and error pipes.
+          It takes a Popen object, a bool indictating whether the data came
+          from the child's stdout (True) or stderr (False), and the bytes
+          buffer to append the data to.
```
Comment (gpshead) [inline Lib/subprocess.py]: it's not "the bytes buffer to append the data to", it is a list of bytes objects to append the new bytes to.

File: Doc/library/subprocess.rst [hunk 11]
```diff
@@ -654,6 +655,17 @@ functions.
    .. versionadded:: 3.10
       The ``pipesize`` parameter was added.
 
+   *read_callback* can be used to replace the default internal handler for
```
Comment (gpshead) [inline Doc/library/subprocess.rst]: I'd rather not mention internal implementation details in the public docs.  Instead, a wording more like:

```
*read_callback*, if supplied, is called upon :data:`PIPE` data being read
for either *stdout* or *stderr* of the child process. It must expect four args:
The :class:``Popen`` instance, an ``is_stdout`` (bool), an ``output_buffer``
list of bytes to append accumulated data to, and the ``data`` (bytes) just read
from the pipe. In order to maintain standard Popen behavior of collecting
pipe data to be available at the finish in *stdout* and *stderr* attributes,
``output_buffer.append(data)`` must be called within the callback.
```

Or something along those lines. It is more friendly. No mention of internal things.

Yet even writing that... the way this reads, it suggests the API for the callback really should be different. The arguments are not supplied in the order they need to be discussed.  How about:

`(self: subprocess.Popen, data: bytes, output_list: list[bytes])`

as the callback API.

I'm still torn on the *is_stdout* bool. It is annoying to see a bool. It means many implementations also need to have a conditional on it.  I think we should have two callbacks instead.  read_stdout_callback & read_stderr_callback.  People who want the same behavior on both just pass both args pointing to the same callable.  People who want different behavior supply different callbacks.

File: Lib/subprocess.py [hunk 12]
```diff
@@ -756,19 +756,37 @@ class Popen:
...
+
+      tee: If true, the output and/or error streams are also written to
+          the parent's sys.stdout and/or sys.stderr, respectively.
+
     Attributes:
         stdin, stdout, stderr, pid, returncode
     """
     _child_created = False  # Set here since __del__ checks it
 
+    def _read_handler(self, is_stdout, buffer, data):
```
Comment (gpshead) [inline Lib/subprocess.py]: class organization style wise, never write a method before the constructor in the class body.

File: Doc/library/subprocess.rst [hunk 13]
```diff
@@ -654,6 +655,17 @@ functions.
...
+   reading the :data:`PIPE` used for *stdout* or *stderr* of the child process.
+   Replacements must accept the same parameters, which are ``self``,
+   ``is_stdout`` (bool), the ``buffer`` (bytes) to accumulate data into,
+   and the ``data`` (bytes) just read from the pipe.
+
+   *tee* when set, indicates that the pipe reader callback should not only
```
Comment (gpshead) [inline Doc/library/subprocess.rst]: This should be described without mentioning a reader callback at all. That is an implementation detail. Rather than a parameter for this, perhaps providing a predefined read callback that implements tee would be better.  combined with splitting the callback into two, this would allow people to tee one or the other easily.

**Review by gpshead — CHANGES_REQUESTED** [review]
this is taking shape. I think I want to change the shape of it a bit. We don't need the tee parameter & we should have two callbacks instead of one that is fed a bool as to which output handle it is being called on. we can provide a pre-written tee callback implementation for the user (and even use its trivial implementation as an example in the docs).

File: Doc/library/subprocess.rst (hunk 11 above)
Comment (pprindeville) [reply Doc/library/subprocess.rst]: Done.

File: Lib/subprocess.py (hunk 10 above)
Comment (pprindeville) [reply Lib/subprocess.py]: Done.

File: Doc/library/subprocess.rst (hunk 13 above)
Comment (pprindeville) [reply Doc/library/subprocess.rst]: I've removed mention of the reader callback.

File: Lib/subprocess.py (hunk 12 above)
Comment (pprindeville) [reply Lib/subprocess.py]: Done.

File: Lib/subprocess.py (hunk 9 above)
Comment (pprindeville) [reply Lib/subprocess.py:1598]: How does one then handle a disparity of rates between how quickly one can read from `fh` and how quickly they can write to `sys.stdout`?  I don't see a simple solution, unless I've misunderstood what you're saying.

**Comment by pprindeville** [comment]
I have made the requested changes; please review again

File: Lib/subprocess.py (hunk 9 above)
Comment (gpshead) [reply Lib/subprocess.py:1598]: I don't expect any `tee` like process to handle this. The posix side of things in this code doesn't either, it reads up to 32k when the selector event says the pipe has data and calls the callback on that.  if the callback blocks, the child process can thus also block on IO when one of its pipe buffers fills up.

there's also the issue of threading, if a callback _might_ be called from a different thread rather than in the thread calling Popen.communicate, that needs to be documented.

Comment (pprindeville) [reply Lib/subprocess.py:1598]: I'm wondering if that's a constraint of the original design that we're just uncovering?  Is that really an issue to fix here, or just something to call out and having been exposed...

Comment (gpshead) [reply Lib/subprocess.py:1598]: The original design consumes all pipe data as it comes in and buffers it as it is only available as a whole upon process exit when `communicate()` returns. The reason I bring this up here is that the tee and callback behavior would be inconsistent between Windows and posix such that the a read callback and tee wouldn't be very useful on Windows.

**Comment by pprindeville** [comment]
> The original design consumes all pipe data as it comes in and buffers it as it is only available as a whole upon process exit when communicate() returns. The reason I bring this up here is that the tee and callback behavior would be inconsistent between Windows and posix such that the a read callback and tee wouldn't be very useful on Windows.

I can add code to disable the `tee` functionality on Windows.

File: Lib/subprocess.py (hunk 9 above)
Comment (pprindeville) [reply Lib/subprocess.py:1598]: I can add code to disable the `tee` functionality on Windows; if Windows users want the functionality parity then they can do a rewrite of how I/O is handled on Windows, maybe using the AsyncIO plumbing...

Comment (pprindeville) [reply Lib/subprocess.py:1598]: @gpshead Do you want me to proceed with that change?

Comment (pprindeville) [reply Lib/subprocess.py:1598]: @gpshead I don't know how to proceed.  It would seem that the functionality of this feature has additional contractual obligations on Windows that aren't implied anywhere else.  Should I just limit the functionality of the baked-in "tee" functionality to non-Windows (i.e. POSIX) platforms instead?

**Comment by pprindeville** [comment]
I have made the requested changes; please review again

File: Lib/subprocess.py (hunk 9 above)
Comment (pprindeville) [reply Lib/subprocess.py:1598]: I've disabled the `tee` functionality for Windows.

**Comment by pprindeville** [comment]
@gpshead @gst Is this DOA?  Should I just close it?

File: Lib/subprocess.py (hunk 9 above)
Comment (gpshead) [reply Lib/subprocess.py:1598]: This still leaves Windows behind, it won't make callbacks in real time as data comes in. It isn't about `tee` so much as it's about having a similar user experience and behavior on both platforms. We need to conditionally do bounded read()s in a loop on windows when a callback is set.

From a user point of view "good enough" is likely be to call `fh.readline(2000)` in a loop in this situation. The things someone wants to see data from live are often text based so reading by line, if present, otherwise a limited buffer is roughly the same to non-blocking reads of everything available. something like this:

```
if not callback:
    buffer.append(fh.read())  # the existing logic
else:
    data = None
    while data != b'':
        data = fh.readline(2000)
        callback(self, buffer, data)
```

**Review by gpshead — CHANGES_REQUESTED** [review]
I still agree with my previous comment saying to remove `tee` as a parameter and instead provide a pre-made set of public tee-providing callbacks in the module which are basically what you have as `_read_stdout_handler` and friend internally today. Something like this in the module:

```
def tee_pipe_to(output_fh):
    def tee_to_pipe(buffer, data):
        buffer.append(data)
        output_fh.write(data)
    return tee_to_pipe
```

Would allow for:

`Popen(..., read_stdout_callback=tee_pipe_to(sys.stdout.buffer))`

To just tee stdout.  If the process were launched with text io, that'd change to `sys.stdout` without the `.buffer` on it.

**Comment by pprindeville** [comment]
>From a user point of view "good enough" is likely be to call fh.readline(2000) in a loop in this situation. The things someone wants to see data from live are often text based so reading by line, if present, otherwise a limited buffer is roughly the same to non-blocking reads of everything available. something like this:
> 
> ```
> if not callback:
>     buffer.append(fh.read())  # the existing logic
> else:
>     data = None
>     while data != b'':
>         data = fh.readline(2000)
>         callback(self, buffer, data)
> ```

For simplicity and consistency, why not just always loop reading 2000 lines at a time?

File: Lib/subprocess.py (hunk 9 above)
Comment (pprindeville) [reply Lib/subprocess.py:1598]: > so much as it's about having a similar user experience and behavior on both platforms

It seems like we're bifurcating the behavior on Windows depending on whether the callback has been installed or not.  Why not just have consistent and predictable behavior?

**Comment by gpshead** [comment]
> For simplicity and consistency, why not just always loop reading 2000 lines at a time?

`.readline(2000)` bounds the read at 2000 bytes, not 2000 lines. Always doing that would be a performance regression for the most common existing users case: Where no checking for EOLs is done and the unbounded read() work is all in C with a more optimal read buffer size and result storage reallocation code.

**Comment by pprindeville** [comment]
> > For simplicity and consistency, why not just always loop reading 2000 lines at a time?
> 
> `.readline(2000)` bounds the read at 2000 bytes, not 2000 lines. Always doing that would be a performance regression for the most common existing users case: Where no checking for EOLs is done and the unbounded read() work is all in C with a more optimal read buffer size and result storage reallocation code.

Okay, I'll work on a "fast path" when the handler isn't overridden.  In the meantime, though, I'm failing CI/CD on Windows but I don't have a Windows machine (VM) to test on.  See anything meaningful in the error logs?

**Comment by pprindeville** [comment]
@gpshead Trying your change with `.readline()`, the test `test_check_output_input_none_text` seems to loop forever.  Not sure why.

It's nearly identical to the preceding test `test_check_output_input_none`, but differs only in adding `text=True`.

**Comment by pprindeville** [comment]
> I'll try and take a look this week.

Thanks. I appreciate all of your efforts.

**Comment by pprindeville** [comment]
> I'll try and take a look this week.

Any luck?

**Comment by pprindeville** [comment]
@gpshead Not sure what to do at this point.

**Comment by gpshead** [comment]
Thanks for the ping. Looking into this is still on my TODO list.

**Comment by pprindeville** [comment]
Should I abandon this?

**Comment by gpshead** [comment]
no, leave it open. i haven't had time.

**Comment by pprindeville** [comment]
Any progress?

**Comment by pprindeville** [comment]
> Let's not go forward with this PR, see my overall comment on the issue. Thanks for your work on it! The particular implementation in progress being done here isn't the reason why, though it is an example of why this complexity is not great to maintain within the stdlib.

I won't be contributing again.

**Comment by gst** [comment]
Hi @pprindeville , this idea would be to me very welcome as an independent normal library package/module.

I personally remember having needed and implemented such kind of feature, but was using thread(s) to consume and share to other thread(s) (and be able to retain the entire output as well) the subprocess output as it arrives basically. but I wished there would have been such package available, would have me saved of that extra work.

I understand your frustration but the invoked reason is really not bad/valid actually.

Anyway, have a good weekend.
